lib/uknofault/maccess.c

```diff
@@ -0,0 +1,125 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/nofault.h>
+#include <uk/arch/types.h>
+#include <uk/arch/traps.h>
+#include <uk/event.h>
+
+#include <arch/maccess.h>
+
+/* TODO: Set if ukvmem is available */
+#define nf_disable_paging() do { } while (0)
+#define nf_enable_paging()  do { } while (0)
+
+/* Must be a library global symbol so that the linker can resolve it */
+int nf_mf_handler(const struct nf_excpttab_entry *e,
+		  struct ukarch_trap_ctx *ctx)
+{
+	/* Just continue execution at the fault label */
+	nf_regs_ip(ctx->regs) = nf_excpttab_get_cont_ip(e);
+
+	return UK_EVENT_HANDLED;
+}
+
+static int nf_mem_fault_handler(void *data)
+{
+	struct ukarch_trap_ctx *ctx = (struct ukarch_trap_ctx *)data;
+
+	return nf_handle_trap(nf_regs_ip(ctx->regs), ctx);
+}
+
+UK_EVENT_HANDLER_PRIO(UKARCH_TRAP_PAGE_FAULT, nf_mem_fault_handler,
+		      CONFIG_LIBUKNOFAULT_FAULT_PRIO);
+UK_EVENT_HANDLER_PRIO(UKARCH_TRAP_BUS_ERROR,  nf_mem_fault_handler,
+		      CONFIG_LIBUKNOFAULT_FAULT_PRIO);
+#ifdef CONFIG_ARCH_X86_64
+UK_EVENT_HANDLER_PRIO(UKARCH_TRAP_X86_GP,     nf_mem_fault_handler,
+		      CONFIG_LIBUKNOFAULT_FAULT_PRIO);
+#endif /* CONFIG_ARCH_X86_64 */
+
+#define nf_probe_r_loop(addr, len, type, cont_label)			\
+	while (len >= sizeof(type)) {					\
+		nf_probe_r((type *)(addr), type, cont_label);		\
+		addr += sizeof(type);					\
+		len  -= sizeof(type);					\
+	}
+
+#define nf_memcpy_loop(dst, src, len, type, cont_label)			\
+	while (len >= sizeof(type)) {					\
+		nf_memcpy((type *)(dst), (type *)(src),			\
+			  type, cont_label);				\
+		dst += sizeof(type);					\
+		src += sizeof(type);					\
+		len -= sizeof(type);					\
+	}
+
+__sz uk_nofault_probe_r(__vaddr_t addr, __sz len)
+{
+	__sz l = len;
+
+	nf_probe_r_loop(addr, l, __u64, FAULT_1);
+	nf_probe_r_loop(addr, l, __u32, FAULT_1);
+	nf_probe_r_loop(addr, l, __u16, FAULT_1);
+FAULT_1:
+	nf_probe_r_loop(addr, l, __u8, FAULT_EXIT);
+FAULT_EXIT:
+	return len - l;
+}
+
+__sz uk_nofault_probe_r_nopage(__vaddr_t addr, __sz len)
+{
+	__sz r;
+
+	nf_disable_paging();
```

Comment:
Disabling interrupts is not necessary. At least at the moment. Nevertheless, I also realized that it would be better to use a _save flags_ / _restore flags_ approach as it might be that paging is already disabled and you don't want to inadvertently enable it. So I changed the code so that we save and restore state. If you would need to also disable interrupts, either the VM implementation will take care or you have the flexibility here to do it, too, now.

Besides a `nopage` version, it is useful to have a version that continues to probe after the first fault. So you can easily ensure that a whole address range is **NOT** readable/writable (useful in tests, for example). 

But adding all the suffixes to the function names is not very practical. So I added a generic flags parameter. Especially, this allows having all the combinations.